### PR TITLE
ONNXToTOSA: Batchnormalization

### PR DESCRIPTION
**Pull request by ge28boj**

### CODE REVIEW

File: test/mlir/conversion/onnx_to_tosa/NN/BatchNorm.mlir [hunk 1]
```diff
@@ -0,0 +1,60 @@
+// RUN: onnx-mlir-opt --shape-inference --convert-onnx-to-tosa -cse %s -split-input-file | FileCheck %s
+
```
Comment (mgehre-amd) [inline test/mlir/conversion/onnx_to_tosa/NN/BatchNorm.mlir:2]: Could you please add one test case for bf16, f16 & f64? Those are all legal types for BatchNorm.

The .td file shows that BatchNormalizationInferenceMode also accepts a momentum attribute. Can you please add a test case with that?

File: src/Conversion/ONNXToTOSA/NN/BatchNorm.cpp [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+                                 Value &result) {
+      auto toBCastType = toBcast.getType().dyn_cast<RankedTensorType>();
+      if (toBCastType.getRank() > 1) {
+        return rewriter.notifyMatchFailure(op, "Rank cannot be more than 1");
+      }
+      auto outTensorType = outType.cast<RankedTensorType>();
+
+      // onnx-mlir uses layout NCHW for input. For batch normalization, the C
+      // dimension is kept. The new shape should be {1, C, 1, 1}, if the rank of
+      // input/output tensor is 4
```
Comment (mgehre-amd) [inline src/Conversion/ONNXToTOSA/NN/BatchNorm.cpp:70]:
```suggestion
      // The layout of the output is N x C x D1 x D2 … Dn. For batch normalization, the C
      // dimension is kept. The new shape should be {1, C, 1, 1, ...}.
```

File: src/Conversion/ONNXToTOSA/NN/BatchNorm.cpp [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+      SmallVector<int64_t> newShape = {1, toBCastType.getShape()[0]};
+      for (auto i = 2; i < outTensorType.getRank(); i++) {
```
Comment (mgehre-amd) [inline src/Conversion/ONNXToTOSA/NN/BatchNorm.cpp:72]: nit: LLVM style is to elide braces when the body is a single line

File: src/Conversion/ONNXToTOSA/NN/BatchNorm.cpp [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+        newShape.push_back(1);
+      }
+
+      auto newType =
```
Comment (mgehre-amd) [inline src/Conversion/ONNXToTOSA/NN/BatchNorm.cpp:76]: Is this the same as outType?

**Review by mgehre-amd — CHANGES_REQUESTED** [review]
Generally looks good.
Would be great if we could extend the testing to cover the whole op specification.

File: src/Conversion/ONNXToTOSA/NN/BatchNorm.cpp [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+
+namespace onnx_mlir {
+
+namespace {
+
+// output = (input - mean) * scale / sqrt(var + eps) + bias
+Value computeBatchNorm(Operation *op, ConversionPatternRewriter &rewriter,
+    Type outType, Value input, Value variance, Value eps, Value mean,
+    Value scale, Value bias) {
+  auto op1SubInputMean =
```
Comment (ttjost) [inline src/Conversion/ONNXToTOSA/NN/BatchNorm.cpp:30]: Could you also make use of `tosabuilder` here?

File: src/Conversion/ONNXToTOSA/NN/BatchNorm.cpp [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+  using OpConversionPattern::OpConversionPattern;
+  using OpAdaptor = typename ONNXBatchNormalizationInferenceModeOp::Adaptor;
+  LogicalResult matchAndRewrite(ONNXBatchNormalizationInferenceModeOp op,
+      OpAdaptor adaptor, ConversionPatternRewriter &rewriter) const override {
+
+    auto outType = getTypeConverter()->convertType(op.getResult().getType());
+
+    // reshape rank-1 tensors (scale, bias, mean, variance) and attribute
+    // epsilon, such that they have the same rank as input/output tensor
+    auto reshapeToInputDim = [&](Operation *op,
```
Comment (ttjost) [inline src/Conversion/ONNXToTOSA/NN/BatchNorm.cpp:58]: If you use tosabuilder, I think this function is not necessary. Everything is handled for free. Let me know if you need to discuss about it.

**Review by ttjost — CHANGES_REQUESTED** [review]
Could you please use `tosabuilder`? That will make things easier.

**Review by mgehre-amd — APPROVED** [review]

File: test/mlir/conversion/onnx_to_tosa/NN/BatchNorm.mlir [hunk 7]
```diff
@@ -18,43 +17,89 @@ func.func @test_batchnorm(%arg0: tensor<100x3x10x10xf32>) -> tensor<100x3x10x10x
...
+    %4 = "onnx.BatchNormalizationInferenceMode"(%arg0, %0, %1, %2, %3) {epsilon = 1.00000007E-5 : f32} : (tensor<100x3x10x10xf64>, tensor<3xf64>, tensor<3xf64>, tensor<3xf64>, tensor<3xf64>) -> tensor<100x3x10x10xf64>
+    return %4 : tensor<100x3x10x10xf64>
+// CHECK-LABEL: func @test_batchnorm_f64
+// CHECK-SAME:  ([[PARAM_0_:%.+]]: tensor<100x3x10x10xf64>) -> tensor<100x3x10x10xf64>
+// CHECK:           [[VAR_0_:%.+]] = onnx.Constant dense<[1.000000e+00, 2.000000e+00, 3.000000e+00]> : tensor<3xf64>
+// CHECK:           [[VAR_1_:%.+]] = onnx.Constant dense<[2.000000e+00, 3.000000e+00, 4.000000e+00]> : tensor<3xf64>
+// CHECK:           [[VAR_2_:%.+]] = onnx.Constant dense<[3.000000e+00, 4.000000e+00, 5.000000e+00]> : tensor<3xf64>
+// CHECK:           [[VAR_3_:%.+]] = onnx.Constant dense<[4.000000e+00, 5.000000e+00, 6.000000e+00]> : tensor<3xf64>
+// CHECK:           [[VAR_4_:%.+]] = "onnx.BatchNormalizationInferenceMode"([[PARAM_0_]], [[VAR_0_]], [[VAR_1_]], [[VAR_2_]], [[VAR_3_]])
+// CHECK:           return [[VAR_4_]] : tensor<100x3x10x10xf64>
```
Comment (roberteg16) [inline test/mlir/conversion/onnx_to_tosa/NN/BatchNorm.mlir:104]: Normally when we want to check that a legalization did not happen it is enough if you check that the op is still there:

```suggestion
//     CHECK: onnx.BatchNormalizationInferenceMode
// CHECK-NOT: tosa
```

File: test/mlir/conversion/onnx_to_tosa/NN/BatchNorm.mlir [hunk 8]
```diff
@@ -18,43 +17,89 @@ func.func @test_batchnorm(%arg0: tensor<100x3x10x10xf32>) -> tensor<100x3x10x10x
...
+// CHECK-LABEL: func @test_batchnorm_bf16_dynamic
+// CHECK-SAME:  ([[PARAM_0_:%.+]]: tensor<100x3x?x?xbf16>) -> tensor<100x3x?x?xbf16>
+// CHECK-DAG:       [[VAR_0_:%.+]] = "tosa.const"() <{value = dense<[1.000000e+00, 2.000000e+00, 3.000000e+00]> : tensor<3xbf16>}> : () -> tensor<3xbf16>
+// CHECK-DAG:       [[VAR_1_:%.+]] = "tosa.const"() <{value = dense<[2.000000e+00, 3.000000e+00, 4.000000e+00]> : tensor<3xbf16>}> : () -> tensor<3xbf16>
+// CHECK-DAG:       [[VAR_2_:%.+]] = "tosa.const"() <{value = dense<[3.000000e+00, 4.000000e+00, 5.000000e+00]> : tensor<3xbf16>}> : () -> tensor<3xbf16>
+// CHECK-DAG:       [[VAR_3_:%.+]] = "tosa.const"() <{value = dense<[4.000000e+00, 5.000000e+00, 6.000000e+00]> : tensor<3xbf16>}> : () -> tensor<3xbf16>
+// CHECK-DAG:       [[VAR_4_:%.+]] = tosa.reshape [[VAR_2_]] {new_shape = array<i64: 1, 3, 1, 1>} : (tensor<3xbf16>) -> tensor<1x3x1x1xbf16>
+// CHECK-DAG:       [[VAR_5_:%.+]] = tosa.reshape [[VAR_0_]] {new_shape = array<i64: 1, 3, 1, 1>} : (tensor<3xbf16>) -> tensor<1x3x1x1xbf16>
+// CHECK-DAG:       [[VAR_6_:%.+]] = tosa.reshape [[VAR_1_]] {new_shape = array<i64: 1, 3, 1, 1>} : (tensor<3xbf16>) -> tensor<1x3x1x1xbf16>
```
Comment (roberteg16) [inline test/mlir/conversion/onnx_to_tosa/NN/BatchNorm.mlir:75]: NIT: Usually we take into consideration the semantics of the op when we write the test. Reshapes can be simplified to:

``` suggestion
// CHECK-DAG:       [[VAR_4_:%.+]] = tosa.reshape [[VAR_2_]] {new_shape = array<i64: 1, 3, 1, 1>} : (tensor<3xbf16>)
// CHECK-DAG:       [[VAR_5_:%.+]] = tosa.reshape [[VAR_0_]] {new_shape = array<i64: 1, 3, 1, 1>} : (tensor<3xbf16>)
// CHECK-DAG:       [[VAR_6_:%.+]] = tosa.reshape [[VAR_1_]] {new_shape = array<i64: 1, 3, 1, 1>} : (tensor<3xbf16>)
```

File: src/Conversion/ONNXToTOSA/NN/BatchNorm.cpp [hunk 9]
```diff
@@ -52,77 +29,53 @@ class ONNXBatchNormalizationInferenceModeOpLoweringToTOSA
...
+    Value reshapedMean;
+    Value reshapedScale;
+    Value reshapedBias;
+    Value reshapedVar;
+
+    reshapedMean = tosaBuilder.reshape(mean, llvm::ArrayRef<int64_t>(newShape));
+    reshapedScale =
+        tosaBuilder.reshape(scale, llvm::ArrayRef<int64_t>(newShape));
+    reshapedBias = tosaBuilder.reshape(bias, llvm::ArrayRef<int64_t>(newShape));
+    reshapedVar = tosaBuilder.reshape(var, llvm::ArrayRef<int64_t>(newShape));
```
Comment (roberteg16) [inline src/Conversion/ONNXToTOSA/NN/BatchNorm.cpp:60]: NIT: I believe you can remove `llvm::`

**Review by roberteg16 — APPROVED** [review]
LGTM

**Review by ttjost — APPROVED** [review]

**Pull request merged**